src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -20,6 +20,10 @@ public sealed partial class JsonSerializerOptions
     {
         internal const int BufferSizeDefault = 16 * 1024;
 
+        // For backward compatibility the default max depth for JsonSerializer is 64,
+        // the minimum of JsonReaderOptions.DefaultMaxDepth and JsonWriterOptions.DefaultMaxDepth.
+        internal const int DefaultMaxDepth = JsonReaderOptions.DefaultMaxDepth;
```

Comment:
Backwards compat would imply 1000 for write (and 64 for read).

Technically, this is a breaking change for those that write JSON > 64 levels but do not read that JSON back; for those that read and write the same JSON, the serializer's MaxDepth (either the default value or the value set explicitly) would be fine.

Reply:
> Backwards compat would imply 1000 for write (and 64 for read).

Currently `JsonSerializer` forces a max depth of 64 when writing (unless a nonzero `JsonSerializerOptions.MaxDepth` setting is specified). This is a [check separate to the `Utf8JsonWriter` layer](https://github.com/dotnet/runtime/blob/2b1f420cca78ca03f867caa68f30cf81ae2b4187/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L323).

Reply:
Ah that's right I forgot the serializer checks for that. Can we now remove that logic and let the writer throw?

Reply:
Good question. The converter check is certainly redundant (and a bit hit-or-miss, see https://github.com/dotnet/runtime/issues/50321#issuecomment-892695769) so removing it would an improvement. The problem though is that the converter throws a `JsonException` with a message indicating the presence of a potential cycle whereas the `Utf8JsonWriter` throws an [`InvalidOperationException`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs#L430) which will get [wrapped in a `JsonException`](https://github.com/dotnet/runtime/blob/0749730dbbe7e9815d99839c3b9333ea409ba98f/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.ReadCore.cs#L89) by the serializer. 

I don't think making that change would be a huge break. cc @layomia for a second opinion.

Reply:
I pushed commit 0e430928c8d84288e0b2cb2e1d1726087f905ae8 that introduces the change. However I feel that from a user experience perspective this is a regression. The error message changes from "Possible object cycle detected" to "The object or value could not be serialized" with an inner exception containing the actual message "CurrentDepth (64) is equal to or larger than the maximum allowed depth of 64. Cannot write the next JSON object or array."

Reply:
Can we modify the `InvalidOperationException` message thrown by the writer to say something like "If called from the serializer, {cycle message}".

Reply:
I do like the 2 `if` statements that were removed from the serializer's write() method.

Reply:
> Can we modify the InvalidOperationException message thrown by the writer to say something like "If called from the serializer, {cycle message}".

I don't believe the writer is able to distinguish this. We'd probably need to handle the exception and change it, but that feels awkward. To be honest I think we should just keep the existing duplication.

Reply:
I reverted https://github.com/dotnet/runtime/pull/61608/commits/0e430928c8d84288e0b2cb2e1d1726087f905ae8, since I wasn't satisfied with the quality of the error messages. We can revisit in a separate PR.